packages/machine-extractor/src/extractAction.ts

```diff
@@ -0,0 +1,411 @@
+export type JsonObject = { [key: string]: JsonValue };
+export type JsonArray = JsonValue[];
+export type JsonEntry = [string, { value: JsonValue; type: JsonEntryType }];
+// Using this type to avoid merging 'expression' and string for JsonEntry.type
+type JsonExpression = 'expression' & { __tag: 'JsonExpression' };
+export type JsonValue =
+  | string
+  | number
+  | boolean
+  | null
+  | JsonArray
+  | JsonObject;
+type JsonEntryType = JsonValue | JsonExpression;
+
+export function extractAssignment(
+  actionNode: ActionNode,
+  fileContent: string,
+): Record<
+  string,
+  {
+    type: JsonEntryType;
+    value: JsonValue;
+  }
+> {
+  const node = actionNode.node;
+  const assignment: Record<
+    string,
+    {
+      type: JsonEntryType;
+      value: JsonValue;
+    }
+  > = {};
+  if (t.isCallExpression(node)) {
+    const assigner = node.arguments[0];
+
+    // assign({})
+    if (t.isObjectExpression(assigner)) {
+      assigner.properties.forEach(prop => {
+        if (t.isObjectProperty(prop)) {
+          if (t.isIdentifier(prop.key)) {
+            /**
+             * assign({prop: () => {}})
+             * assign({prop: function() {}})
+             */
+            if (
+              t.isArrowFunctionExpression(prop.value) ||
+              t.isFunctionExpression(prop.value)
+            ) {
+              assignment[prop.key.name] = {
+                type: 'expression',
+                value: fileContent.slice(prop.value.start!, prop.value.end!),
+              };
+            } else if (
+              /**
+               * assign({prop: []})
+               * assign({prop: {}})
+               */
+              t.isArrayExpression(prop.value) ||
+              t.isObjectExpression(prop.value)
+            ) {
+              assignment[prop.key.name] = {
+                type: 'expression',
+                value: fileContent.slice(prop.value.start!, prop.value.end!),
+              };
+            } else if (t.isLiteral(prop.value)) {
+              /**
+               * assign({prop: literal value})
+               */
+              if (
+                t.isRegExpLiteral(prop.value) ||
+                t.isTemplateLiteral(prop.value) ||
+                t.isNullLiteral(prop.value)
+              ) {
+                assignment[prop.key.name] = {
+                  type: 'expression',
+                  value: fileContent.slice(prop.value.start!, prop.value.end!),
+                };
+              } else {
+                assignment[prop.key.name] = {
+                  type: getLiteralType(prop.value),
+                  value: prop.value.value,
+                };
+              }
+            }
+          }
+        }
+      });
+
+      return assignment;
+    }
+
+    // assign(() => {})
+    // assign(function() {})
+    if (
+      t.isArrowFunctionExpression(assigner) ||
+      t.isFunctionExpression(assigner)
+    ) {
+      return {
+        inlineImplementation: {
```

Comment:
I'd replace this with smth like `{ type: 'expression', value: string } | { type: 'object': properties: Record<string, ValidStuff> }`. This way `inlineImplementation` wouldn't have to be a special key here.